proposals/context-parameters.md

```diff
@@ -0,0 +1,571 @@
+
+```kotlin
+context(logger: Logger) fun logWithTime(message: String) =
+  logger.log("${LocalDateTime.now()}: $message")
+
+context(logger: Logger) fun User.doAction() {
+  logWithTime("saving user $id")
+  // ...
+}
+```
+
+**§1.5** *(override)*: Context parameters are part of the signature, and follow the same rules as regular value parameters concerning overriding:
+
+* The type and order of context parameters must coincide.
+* It is allowed (yet discouraged) to change the name of a context parameter.
+
+**§1.6** *(naming ambiguity)*: We use the term **context** with two meanings:
+
+1. For a declaration, it refers to the collection of context parameters declared in its signature. We also use the term *contextual function or property*.
+2. Within a block, we use context to refer to the combination of implicit receivers and context parameters in scope in that block. This context is the source for context resolution, that is, for "filling in" the implicit context parameters.
+
+**§1.7** *(function types)*: **Function types** are extended with context parameters. It is only allowed to mention the *type* of context parameters, names are not supported.
+
+* We do not want to inspect the body of a lambda looking for different context parameter names during overload resolution. The reasoning is similar to how we restrict lambdas with no declared arguments to be 0 or 1-ary.
+
+```kotlin
+context(Transaction) (UserId) -> User?
+context(Logger) User.() -> Int
+```
+
+Note that, like in the case of extension receivers, those types are considered equivalent (for typing purposes, **not** for resolution purposes) to the function types in which all parameters are declared as value parameters _in the same order_.
+
+```kotlin
+// these are all equivalent types
+context(Logger, User) () -> Int
+context(Logger) User.() -> Int
+context(Logger) (User) -> Int
+Logger.(User) -> Int
+(Logger, User) -> Int
+```
+
+**§1.8** *(lambdas)*: If a lambda is assigned a function type with context parameters, those behave as if declared with `_` as its name.
+
+* They participate in context resolution but are only accessible through the `implicit` function (defined below).
+
+```kotlin
+fun <A> withConsoleLogger(block: context(Logger) () -> A) = ...
+
+withConsoleLogger {
+  // you can call functions with Logger as context parameter
+  logWithTime("doing something")
+  // you can use 'implicit' to access the context parameter
+  implicit<Logger>().log("hello")
+}
+```
+
+## Standard library support
+
+**§2.1** *(`context` function)*: The `context` function adds a new value to the context, in an anonymous manner.
+
+* The implementation may be built into the compiler, instead of having a plethora of functions defined in the standard library.
```

Comment:
This likely would complicate providing documentation for them.

Reply:
We might need to be a bit more creative here. Maybe the regular Dokka-style docs are not good in this case; I would imagine this being part of the language documentation, given how central they are going to be to the context parameters feature itself.